### PR TITLE
[SPARK-6991] [SparkR] Adds support for zipPartitions.

### PR DESCRIPTION
**Pull request by hlin09**

### CODE REVIEW

**Comment by concretevitamin** [comment]
Thanks @hlin09 - one high-level question: do you have an intended use case in mind for arbitrary / large arity functions? In Spark's Scala API up to 4 RDDs are supported, which has perf. benefits (avoid an union + shuffle).

**Comment by hlin09** [comment]
I do have some user cases that would best to have more that 4 arity zip function. Another concern is that the current communication pattern between Spark and R does not allows tuples of more than RDDs. From perf point of view, "union + shuffle" might be better than recursively zip 2 of them.

File: R/pkg/R/RDD.R [hunk 1]
```diff
@@ -1529,3 +1529,50 @@ setMethod("zipRDD",
             
             PipelinedRDD(zippedRDD, partitionFunc)
           })
+
+#' Zip an RDD's partitions with one (or more) RDD(s).
+#'
+#' Return a new RDD by applying a function to the zipped partitions. 
+#' Assumes that all the RDDs have the same number of partitions*, but does *not*
```
Comment (concretevitamin) [inline R/pkg/R/RDD.R]: nit: unmatched *

File: R/pkg/R/RDD.R [hunk 2]
```diff
@@ -1529,3 +1529,50 @@ setMethod("zipRDD",
...
+            if (length(rrdds) == 1) {
+              return(rrdds[[1]])
+            }
+            nPart <- sapply(rrdds, numPartitions)
+            if (length(unique(nPart)) != 1) {
+              stop("Can only zipPartitions RDDs which have the same number of partitions.")
+            }
+            
+            rrdds <- lapply(rrdds, function(rdd) {
+              mapPartitionsWithIndex(rdd, function(split, part) {
```
Comment (concretevitamin) [inline R/pkg/R/RDD.R]: We've recently changed all usage of `split` to `partIndex` or something like `idx` is fine too.

File: R/pkg/R/RDD.R [hunk 3]
```diff
@@ -1529,3 +1529,50 @@ setMethod("zipRDD",
...
           })
+
+#' Zip an RDD's partitions with one (or more) RDD(s).
+#'
+#' Return a new RDD by applying a function to the zipped partitions. 
+#' Assumes that all the RDDs have the same number of partitions*, but does *not* 
+#' require them to have the same number of elements in each partition.
+#' 
+#' @param x An RDD to be zipped.
+#' @param other Another RDD to be zipped.
```
Comment (concretevitamin) [inline R/pkg/R/RDD.R]: The doc here needs to be updated: (1) var-arg arity (2) document `func`.

**Comment by concretevitamin** [comment]
@hlin09 - cool, I don't have a strong opinion towards either approach for now. Did a pass and left some minor style comments.

File: R/pkg/R/RDD.R (hunk 1 above)
Comment (hlin09) [inline R/pkg/R/RDD.R]: Thanks. Done.

File: R/pkg/R/RDD.R (hunk 2 above)
Comment (hlin09) [inline R/pkg/R/RDD.R]: Done.

File: R/pkg/R/RDD.R (hunk 3 above)
Comment (hlin09) [inline R/pkg/R/RDD.R]: Done.

**Comment by concretevitamin** [comment]
LGTM. /cc @shivaram

**Comment by shivaram** [comment]
Thanks @concretevitamin for the review. Will merge this after Jenkins passes

**Comment by shivaram** [comment]
Hmm this is weird - somehow the test results for this PR were never posted to github. https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31005/consoleFull reports a json parsing error

cc @shaneknapp

**Comment by shivaram** [comment]
I also re-ran the R tests locally to just verify -- Merging this.

**Pull request closed without merging**